src/framework/constraints.ml

```diff
@@ -635,19 +669,165 @@ struct
+    let longjmppaths = List.filter (fun (c,fc,v) -> not (D.is_bot v)) longjmppaths in
+    let longjmppaths = List.map (Tuple3.map2 Option.some) longjmppaths in
+    let _ = List.iter handlelongjmp longjmppaths in
+    (* Return result of normal call, longjmp om;y happens via side-effect *)
+    result
+
+  let tf_special_call ctx (getl: lv -> ld) (sidel: lv -> ld -> unit) lv f args =
+    let jmptarget = function
+      | Statement s -> LongjmpTo s
+      | _ -> failwith "should not happen"
+    in
+    match (LibraryFunctions.find f).special args with
+    | Setjmp { env; savesigs} ->
+      (* Handling of returning for the first time *)
+      let first_return = S.special ctx lv f args in
+      if M.tracing then Messages.tracel "longjmp" "reading from %s\n" (Node.show (jmptarget ctx.prev_node));
+      let later_return = getl (jmptarget ctx.prev_node, ctx.context ()) in
+      if not @@ S.D.is_bot later_return then
+        S.D.join first_return later_return
+      else
+        first_return
+    | Longjmp {env; value; sigrestore} ->
+      let res = S.special ctx lv f args in
+      let current_fundec = Node.find_fundec ctx.node in
+      let one_path s = (
+        let rec path_ctx = { ctx with
+                             ask = (fun (type a) (q: a Queries.t) -> S.query path_ctx q);
+                             local = s;
+                           }
+        in
+        (* Eval `env` again to avoid having to construct bespoke ctx to ask *)
+        let targets = path_ctx.ask (EvalJumpBuf env) in
+        if M.tracing then Messages.tracel "longjmp" "Jumping to %a\n" JmpBufDomain.JmpBufSet.pretty targets;
+        let handle_longjmp = function
+          | JmpBufDomain.BufferEntryOrTop.AllTargets ->
+            M.warn "Longjmp to potentially invalid target, as contents of buffer %a may be unknown! (imprecision due to heap?)"  d_exp env
+          | JmpBufDomain.BufferEntryOrTop.Target (node, c) ->
+            let validBuffers = path_ctx.ask ValidLongJmp in
+            if not (JmpBufDomain.JmpBufSet.mem (Target (node,c)) validBuffers) then
+              M.warn "Longjmp to potentially invalid target! (Target %s in Function %s which may have already returned or is in a different thread)" (Node.show node) (Node.find_fundec node).svar.vname
+            else
+              (let controlctx = ControlSpecC.hash (ctx.control_context ()) in
+               if c = IntDomain.Flattened.of_int (Int64.of_int controlctx) && (Node.find_fundec node).svar.vname = current_fundec.svar.vname then
+                 (if M.tracing then Messages.tracel "longjmp" "Potentially from same context, side-effect to %s\n" (Node.show node);
+                  match node with
+                  | Statement { skind = Instr [Call (lval, exp, args,_, _)] ;_ } ->
+                    let res' = match lval with
+                      | Some lv -> Goblintutil.assign_is_setjmp := true; let r = S.assign path_ctx lv value in Goblintutil.assign_is_setjmp := false; r
```

Comment:
The resetting needs to be under `Fun.protect`

Reply:
I was able to completely remove `assign_is_setjmp` now.